pkg/template/servicebroker/provision.go

```diff
@@ -0,0 +1,188 @@
+	uuidMap := &api.TemplateBrokerUUIDMap{
+		ObjectMeta: kapi.ObjectMeta{Name: instance_id},
+		Spec: api.TemplateBrokerUUIDMapSpec{
+			TemplateInstanceRef: kapi.ObjectReference{
+				Kind:      "TemplateInstance",
+				Namespace: ti.Namespace,
+				Name:      ti.Name,
+			},
+		},
+	}
+
+	newTemplateBrokerUUIDMap, err := b.oc.TemplateBrokerUUIDMaps().Create(uuidMap)
+	if err == nil {
+		*didWork = true
+		return newTemplateBrokerUUIDMap, nil
+	}
+
+	if kerrors.IsAlreadyExists(err) {
+		var existingUUIDMap *api.TemplateBrokerUUIDMap
+		existingUUIDMap, err = b.oc.TemplateBrokerUUIDMaps().Get(uuidMap.Name)
+		if err == nil && reflect.DeepEqual(uuidMap.Spec, existingUUIDMap.Spec) {
+			return existingUUIDMap, nil
+		}
+
+		return nil, openservicebroker.NewResponse(http.StatusConflict, openservicebroker.ProvisionResponse{}, nil)
+	}
+
+	return nil, openservicebroker.NewResponse(http.StatusInternalServerError, nil, err)
+}
+
+func (b *Broker) Provision(instance_id string, preq *openservicebroker.ProvisionRequest) *openservicebroker.Response {
+	if len(preq.Parameters[api.NamespaceParameterKey]) == 0 {
+		return openservicebroker.NewResponse(http.StatusBadRequest, nil, errors.New("parameters."+api.NamespaceParameterKey+": Required value"))
+	}
+	if validation.ValidateNamespaceName(preq.Parameters[api.NamespaceParameterKey], false) != nil {
+		return openservicebroker.NewResponse(http.StatusBadRequest, nil, errors.New("parameters."+api.NamespaceParameterKey+": Invalid value: must be a valid namespace"))
+	}
+	namespace := preq.Parameters[api.NamespaceParameterKey]
+
+	if len(preq.Parameters[api.ImpersonateUserParameterKey]) == 0 {
+		return openservicebroker.NewResponse(http.StatusBadRequest, nil, errors.New("parameters."+api.ImpersonateUserParameterKey+": Required value"))
+	}
+	if uservalidation.ValidateUserName(preq.Parameters[api.ImpersonateUserParameterKey], false) != nil {
+		return openservicebroker.NewResponse(http.StatusBadRequest, nil, errors.New("parameters."+api.ImpersonateUserParameterKey+": Invalid value: must be a valid username"))
+	}
+	impersonate := preq.Parameters[api.ImpersonateUserParameterKey]
+
+	template, err := b.lister.GetTemplateByUID(preq.ServiceID)
+	if err != nil {
+		return openservicebroker.NewResponse(http.StatusBadRequest, nil, err)
+	}
+	if template == nil {
+		return openservicebroker.NewResponse(http.StatusBadRequest, nil, kerrors.NewNotFound(api.Resource("templates"), preq.ServiceID))
+	}
+
+	if template.ObjectLabels == nil {
+		template.ObjectLabels = make(map[string]string)
+	}
+	template.ObjectLabels[api.TemplateInstanceLabel] = instance_id
+
+	didWork := false
+
+	secret, resp := b.ensureSecret(namespace, instance_id, preq, &didWork)
+	if resp != nil {
+		return resp
+	}
+
+	ti, resp := b.ensureTP(namespace, instance_id, template, secret, impersonate, &didWork)
```

Comment:
@deads2k this is where the TemplateInstance object is assembled and sent to the API to be created.